[Improvement] Draw markers faster

ahuarte47 opened the pull request:
The code takes advantage of points already created to paint the geometries to draw the markers.

m-kuhn commented:
Looks good to me from a logic perspective.

Could you add some more linebreaks to increase readability (after if, else, for...)?

What do you think about using the `Q_FOREACH` macro instead of for loops and pointer arithmetic / random access by using `at( i )`? Are you afraid of a performance impact?

ahuarte47 commented:
I changed it, thanks @m-kuhn ! 

About the Q_FOREACH macro, IMHO the performance impact should be minimum compared to time used by the full rendering task.

nyalldawson commented on src/core/symbology-ng/qgssymbolv2.cpp:
```diff
@@ -785,6 +798,16 @@ void QgsSymbolV2::renderFeature( const QgsFeature& feature, QgsRenderContext& co
+      {
+        markers = pts;
+
+        Q_FOREACH ( QPolygonF hole, holes )
```
You'll want to use `const QPolygonF&` here, will be a bit faster

ahuarte47 replied:
Changed, thanks a lot!

m-kuhn commented:
Waiting for travis to be ok and then it's ready for a public testdrive.

Thanks @ahuarte47

ahuarte47 commented:
Ummm, Travis throws 9 build errors, There is other reasons ?

nyalldawson commented on src/core/symbology-ng/qgssymbolv2.cpp:
```diff
@@ -903,27 +965,9 @@ void QgsSymbolV2::renderFeature( const QgsFeature& feature, QgsRenderContext& co
-    double x, y, z;
-    QPointF mapPoint;
-    while ( geom->geometry()->nextVertex( vertexId, vertexPoint ) )
+    Q_FOREACH ( const QPointF& marker, markers )
```
@ahuarte47 Note that for this one, you actually want a copy rather then the reference. QPointF is small and it's quicker to copy than resolve a reference. So this should be `Q_FOREACH( QPointF marker, markers )` (this rule basically only applies to small inbuilt types like int, double, etc AND some very small Qt objects like QPoint, QPointF, QSize and QSizeF)

nyalldawson commented:
@ahuarte47 really nice! This also has the benefit of using the clipped geometry to draw markers, instead of drawing a bunch of markers which aren't visible in the current map extent.

Those build failures are unrelated - it's a clang error which we've seen once before. Let me test something and I'll get back to you on this.

ahuarte47 committed:
Draw markers faster

The code takes advantage of points already created to paint the
geometries to draw the markers.

ahuarte47 commented:
Hy Nyall, Thanks alot for all support!, all days I learn from your great knowledge!

nyalldawson commented:
@ahuarte47 can you try rebasing against master now?

ahuarte47 commented:
Sure!

m-kuhn commented:
Thanks a lot!

Will you be looking into tinywkb next? ;)

ahuarte47 commented:
Thanks a lot Mathias and Nyall for your support!

Now, I am developing a new pull about [simplification](https://github.com/qgis/QGIS/pull/2744#issuecomment-190717940).

But I gets a big performance issue using QgsGeos in [QgsGeometry](https://github.com/qgis/QGIS/pull/2859)
